### PR TITLE
Add an option to filter by genotype fields in GATK VariantFiltration

### PR DESCRIPTION
**Pull request by OtonicarJan**
## Checklist

* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [x] All inputs are used in process.
* [x] All output fields have a value assigned to them.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of the tool should be self explanatory with links to the tool's
  homepage and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required = False`` is not needed.
* Add descriptive error and warning messages.
* Use ``Cmd()`` for each call in the process (call of a tool, a
  Bash command, a Python script or an R script) and handle return codes.
* Add ``self.progress()`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files:
  * Genomes: ``genome_species.fa.gz``
  * Annotations: ``annotation_species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Specific files should be saved in a folder with the name of the tested tool.
  This folder should include separate subfolders for inputs and outputs.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if the length of some list is 0, if a string is given where one
  expects an integer, if a file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/processes/variant_calling/gatk_variant_filtration.py [hunk 1]
```diff
@@ -83,6 +83,29 @@ class Input:
...
+            "To filter by alternative allele depth, use the expression: "
+            "'vc.getGenotype('sample_name').getAD().1 < n",
+            required=False,
+        )
+        genotype_filter_name = ListField(
+            StringField(),
+            label="Names to use for the list of genotype filters",
+            description="Similar to the INFO field based expressions, but used on the FORMAT "
+            "(genotype) fields instead. Warning: filter names should be in the same order as "
+            "filter expressions. ",
```
Comment (AnzeLovse) [inline resolwe_bio/processes/variant_calling/gatk_variant_filtration.py:106]: Remove space after period.

File: resolwe_bio/processes/variant_calling/gatk_variant_filtration.py [hunk 2]
```diff
@@ -83,6 +83,29 @@ class Input:
...
+            description="Similar to the INFO field based expressions, but used on the FORMAT "
+            "(genotype) fields instead. VariantFiltration will add the sample-level FT tag to "
+            "the FORMAT field of filtered samples (this does not affect the record's FILTER tag). "
+            "One can filter normally based on most fields (e.g. 'GQ < 5.0'), but the GT "
+            "(genotype) field is an exception. We have put in convenience methods so that "
+            "one can now filter out hets ('isHet == 1'), refs ('isHomRef == 1'), or homs "
+            "('isHomVar == 1'). Also available are expressions isCalled, isNoCall, isMixed, "
+            "and isAvailable, in accordance with the methods of the Genotype object."
+            "To filter by alternative allele depth, use the expression: "
+            "'vc.getGenotype('sample_name').getAD().1 < n",
```
Comment (AnzeLovse) [inline resolwe_bio/processes/variant_calling/gatk_variant_filtration.py:98]: Perhaps add what `sample_name` and `n` stand for `vc.getGenotype('sample_name').getAD().1 < n`. 

Also the example should end with `'` and the description with a period.

File: resolwe_bio/processes/variant_calling/gatk_variant_filtration.py [hunk 3]
```diff
@@ -246,6 +289,29 @@ class Input:
...
```
Comment (AnzeLovse) [inline resolwe_bio/processes/variant_calling/gatk_variant_filtration.py:304]: Same as above.

File: resolwe_bio/processes/variant_calling/gatk_variant_filtration.py [hunk 4]
```diff
@@ -246,6 +289,29 @@ class Input:
...
+            required=False,
+        )
+        genotype_filter_name = ListField(
+            StringField(),
+            label="Names to use for the list of genotype filters",
+            description="Similar to the INFO field based expressions, but used on the FORMAT "
+            "(genotype) fields instead. Warning: filter names should be in the same order as "
+            "filter expressions. ",
```
Comment (AnzeLovse) [inline resolwe_bio/processes/variant_calling/gatk_variant_filtration.py:312]: Same as above.

File: resolwe_bio/processes/variant_calling/gatk_variant_filtration.py [hunk 5]
```diff
@@ -83,6 +83,29 @@ class Input:
...
+            StringField(),
+            label="Expressions used with FORMAT field to filter",
+            description="Similar to the INFO field based expressions, but used on the FORMAT "
+            "(genotype) fields instead. VariantFiltration will add the sample-level FT tag to "
+            "the FORMAT field of filtered samples (this does not affect the record's FILTER tag). "
+            "One can filter normally based on most fields (e.g. 'GQ < 5.0'), but the GT "
+            "(genotype) field is an exception. We have put in convenience methods so that "
+            "one can now filter out hets ('isHet == 1'), refs ('isHomRef == 1'), or homs "
+            "('isHomVar == 1'). Also available are expressions isCalled, isNoCall, isMixed, "
+            "and isAvailable, in accordance with the methods of the Genotype object."
```
Comment (AnzeLovse) [inline resolwe_bio/processes/variant_calling/gatk_variant_filtration.py:96]: Missing whitespace after period at the end.

File: resolwe_bio/processes/variant_calling/gatk_variant_filtration.py [hunk 6]
```diff
@@ -246,6 +289,29 @@ class Input:
...
```
Comment (AnzeLovse) [inline resolwe_bio/processes/variant_calling/gatk_variant_filtration.py:302]: Same as above.

File: resolwe_bio/tests/processes/test_variant_calling.py [hunk 7]
```diff
@@ -604,6 +623,23 @@ def test_variant_filtration(self):
             compression="gzip",
         )
 
+        filtering_multi_sample = self.run_process(
+            "gatk-variant-filtration",
+            {
+                "vcf": vcf_multi_sample.id,
+                "ref_seq": ref_seq.id,
+                "genotype_filter_expressions": ["AD.1 < 5"],
```
Comment (AnzeLovse) [inline resolwe_bio/tests/processes/test_variant_calling.py:631]: Is this an alternative to `vc.getGenotype('sample_name').getAD().1 < n`?

Comment (AnzeLovse) [reply resolwe_bio/tests/processes/test_variant_calling.py:631]: Thank you, `AD.1 < 5` seems more general.

**Review by jkokosar — APPROVED** [review]

**Pull request merged**